### ISSUE TITLE
compiler: MLA decode op family (DeepSeek-V4 latent attention)

### ISSUE THREAD

**Issue by xzyaoi** [opentela-ai/vkernels#95]
## Context
floe `deepseek_v4/forward.py::DeepseekV4Attention` (+ `forward_step`): shared-KV MQA (`num_key_value_heads=1`) over a **latent cache**: `q_a_proj → q_a_norm → q_b_proj`, partial trailing-slice **interleaved** RoPE applied only to the rope slice of q/latent-k, per-head learnable attention **sink**, a **conjugate RoPE** on the attention output, and a grouped output projection (`DeepseekV4GroupedLinear`).

## Decomposition decision needed
Mostly expressible with existing ops (`linear`, `rms_norm`, partial-rope after #88) plus new kinds:
1. `mla_scores`: latent-K dot + per-head sink column (+ sliding-window branch bound) — one fused task per (b, head) over the latent prefix
2. `conjugate_rope`: output-side rotation (inverse tables) — small elementwise/rope variant
3. grouped `o_proj` (per-head block-diagonal output width) — a `linear` attribute
The interleaved (GPT-J style) rope slice convention is a third rope variant to fold into #88.

## Work
new op kinds + lowerings + reference bodies; decode-path only (prefill chunking out of scope). Latent cache = external pools with per-row valid lengths (#93) and slot indirection (#94).

## Validation
floe `deepseek_v4/forward.py` prefill-vs-decode self-consistency test as the oracle (the package’s own CPU test).

Labels: arch:ops, model:deepseek-v4

## Implementation sketch

**Backend**: Triton, per-(b,h) task shaped like `_h_scores`/`_h_values_gate` (fused sink column), reading the shared latent pool via slot indirection (#94). On gfx942 the comparator is FlyDSL/AITER `mla_fwd_decode` (see `docs/performance/dsa/gfx942.md`) — same math, different emitter; the IR op is backend-agnostic.

**Arithmetic** (decode step, per head h; window keys ∪ compressed entries, all bf16, fp32 accumulation):
```
q   = q_b_proj(rms(q_a_proj(x)))          # low-rank query; q_b_norm = unweighted RMS
q   = rope_slice(q, cos_q, sin_q)          # trailing rope_dim slice only (convention per DeepseekV4Rotary — pin at impl)
kv  = rms(kv_proj(x));  kv = rope_slice(kv, at its own absolute position)   # K == V, 1 shared head
s   = (q · kv_jᵀ) · D^−½  over j ∈ {window keys (|q−j| < W, j ≤ q), selected compressed entries}
s̃  = concat(s, sink_h)                    # learnable per-head scalar logit
p   = softmax(s̃) ;  o = Σ_{j<T} p_j · kv_j   # sink mass absorbs, contributes no value
o   = rope_slice(o, cos_q, −sin_q)         # CONJUGATE rope at the query position
y   = o_b_proj(o_a_proj(o))                # o_a: block-diagonal over o_groups; o_b: dense mix
```
**Dtypes**: activations/latent pool bf16, softmax + conjugate rotation fp32, sinks fp32 scalars `[H]`, logits out bf16. Sliding-window mask computed from absolute positions (i32), not from cache offsets.

**Comment by xzyaoi** [comment]
**MLA decode family shipped compiler-side** — branch `issue-95-mla-decode` (`797e38a`), PR to follow against `main`.

The vertical, per this issue's contract:

- **`mla_scores`** — fused scores + softmax + sink in ONE task per (batch, head): shared-KV MQA dot against the latent pool through the #94-style `Region.indirect` slot gather, sliding-window bound `t ∈ (p−W, p]` (i.e. `|q−t| < W, t ≤ q`), compressed candidates via the **#97** `comp_idx` indirection (+ normalized `block_bias`), and the per-head learnable sink as the LAST column. fp32 accumulation over D, two-pass max-subtract softmax, invalid candidates exact 0.0 (§4.3).
- **`mla_values`** — the context gather as a sibling task: window rows via the slot table, compressed rows via `comp_idx`, sink column **absorbs softmax mass but contributes no value** (tested both ways: all-mass-on-sink → context ≈ 0; dead sink ≡ candidate-only softmax). fp32 accumulate → single store (fp32 default, explicit bf16 out supported and tested).
- **`conjugate_rope`** — output-side rotation by the NEGATIVE angle, exact inverse of the q/k rotation at the same position: `rope → conjugate_rope` round-trip is the identity (fp64 task math ~1e-12; observable through fp32 storage ~1e-6, asserted). Interleaved GPT-J pair convention is PINNED (pairs (2i, 2i+1), tables indexed by pair index, `[rot, D)` passthrough) with a hand-computed exact test. Decode rotates ONLY the query — cached latent rows stay unrotated.
- **Grouped `linear`** — DeepSeek-V4 `GroupedLinear` as a `grouped_heads=H` attribute on the existing `linear` op: block-diagonal per-head projection, read regions restricted to the diagonal blocks so **off-block weight storage is never observed** (NaN-canary tested), 3D `[B, H, K_g]` activations flatten row-major.

Verified: full-chain executor vs independent fp64 oracle (<1e-5), window geometry at row start (`p < W−1`) and row end, #93 per-row ragged positions (p = 0, 5, 31 in one batch), MQA broadcast (H heads, one latent pool), #97 bias affecting only compressed logits, NaN canaries, single-launch accounting (1 launch, 5 phases, B·H tasks per MLA family), device-template CPU mirrors. Suite: **34 passed, 1 skipped** (CUDA-gated device-template test, same flagged gap as #88/#113/#114/#115/#117) — full lane green except the `test_discovery` inventory failure, which reproduces on **plain `origin/main`** (pre-existing; `2ed4a8a`'s contract update doesn't cover the current drift).

Stacked deps: #88 (rope family) and #97 (indexer top-k / block_bias) are ancestors of `main`. Interface note for the in-flight #96: the compressed-entry consumer here is exactly the #97 producer of record — `comp_idx[B, K]` (−1-padded) + optional f32 `block_bias[B, K]`, consumed post-normalization; genuine mismatches become follow-ups per the recorded ruling.

**Comment by xzyaoi** [comment]
Superseded by #129 (Stage 2 megakernel, merged): the MLA decode family (mla_scores / mla_values / conjugate_rope / grouped linear) landed with the DeepSeek-V4 decode step. OP_MLA_SCORES / OP_MLA_VALUES present in operator_ir.py on main. Closing per the supersession note on PR #127.
